src/ast/passes/semantic_analyser.cpp

```diff
@@ -2351,6 +2351,18 @@ void SemanticAnalyser::visit(Cast &cast)
       return;
     }
 
+    if (cast.type.GetNumElements() == 0)
+    {
+      if (cast.type.GetElementTy()->GetSize() == 0)
+        LOG(ERROR, cast.loc, err_) << "Could not determine size of the array";
+      else
+      {
+        // cast to unsized array (e.g. int8[]), determine size from RHS
+        auto num_elems = rhs.GetSize() / cast.type.GetElementTy()->GetSize();
+        cast.type = CreateArray(num_elems, *cast.type.GetElementTy());
```

Comment:
This relies on the later size check right? Cuz otherwise we need to check if the array elem size can divide the cast-ee size correctly.

Reply:
It does, but an explicit check here wouldn't hurt and would make things easier to understand.

Reply:
Hm, I just realized that this can't really happen as we only support integers and integer arrays. The only error we can get is when the array element size is larger than the cast-ee size.

Still, I added the check to get a better error message and capture these cases if we add more types in future.

Reply:
Ah right. All powers of 2 :)